cairo/processTransfer.cairo

```diff
@@ -0,0 +1,180 @@
+    
+    # populate python namespace with input data
+    outcomeData = program_input['initialOutcome']
+    exitRequest = program_input['exitRequest']
+    holdings = program_input['initialHoldings']
+    payouts = []
+
+    # inform proover of the number of assets in the exit
+    ids.assetCount = len(outcomeData)
+    
+    # write the initialOutcome cairo struct
+    writeOutcomeData(outcomeData, ids.initialOutcome.address_)
+
+    # write the initial holdings
+    init_holdings_ptr = ids.initialHoldings
+    for i, holding in enumerate(holdings):
+        memory[init_holdings_ptr + i] = holdings[i]
+    
+    ###
+    # Calculate the funded payouts
+    ###
+
+    # i indexes assets
+    for i in range(len(exitRequest)):
+        # j indexes payout participants array,
+        # p is the participant's order in the allocation
+        for j, p in enumerate(exitRequest[i]):
+            availableToP = holdings[i]
+            for k in range(p):
+                availableToP -= outcomeData[i]['allocations'][k]['amount']
+            if availableToP > 0:
+                payout = min(outcomeData[i]['allocations'][p]['amount'], availableToP)
+                payouts.append({
+                    'asset': outcomeData[i]['asset'],
+                    'destination': outcomeData[i]['allocations'][p]['destination'],
+                    'amount': payout
+                })
+                holdings[i] -= payout
+                outcomeData[i]['allocations'][p]['amount'] -= payout
+    
+    # write the finalOutcome cairo struct
+    writeOutcomeData(outcomeData, ids.finalOutcome.address_)
+
+    # write the payouts struct
+    payouts_ptr_address = ids.payouts.address_
+
+    for i, payout in enumerate(payouts):
+        print('Payout[' + str(i) + ']: ' + str(payout))
+        relative_address = payouts_ptr_address + i * ids.Payout.SIZE
+        memory[relative_address + ids.Payout.asset] = payout['asset']
+        memory[relative_address + ids.Payout.destination] = payout['destination']
+        memory[relative_address + ids.Payout.amount] = payout['asset']
+
+    # write the final holdings
+    holdings_ptr = ids.finalHoldings
+    for i, holding in enumerate(holdings):
+        memory[holdings_ptr + i] = holdings[i]
+    %}
+
+    return (initialOutcome,
+            finalOutcome,
+            payouts,
+            assetCount,
+            initialHoldings,
+            finalHoldings)
+end
+
+func main{output_ptr: felt*, range_check_ptr}():
+    alloc_locals
+    let (local __fp__, _) = get_fp_and_pc()
+
+    # load the transfer params and compute new outcome
+    let (initialOutcome: SingleAssetExit*,
+        finalOutcome: SingleAssetExit*,
+        payouts: Payout*,
+        assetCount: felt,
+        initialHoldings: felt*,
+        finalHoldings: felt*) = process_transfer()
+
+    # validate the transfer
+    assert_valid_holdings_transform(initialHoldings, finalHoldings, assetCount)
+    return ()
```

Comment:
I notice from the GH action that the execution has no 'program output'. Is that because of this line? Should the function return the `finalOutcome` and `payouts`?

Reply:
In short: this version of the program has no output and its validation is via sanity check assertions. Invalid withdrawal requests will fail to run.

To do here is that:
- the calling code provides its own calculated payout structure which is verified against the python calculated payout
- the `SingleAssetExit` data structure requires a flattening function to become compatible with cairo's pre & post merklization utilities